### PR TITLE
Add VectorizedOapRecordReader for Parquet full scan

### PR DESCRIPTION
**Pull request by LuciferYang**
## What changes were proposed in this pull request?

#536 task 4

Add VectorizedOapRecordReader as VectorizedParquetRecordReader to optimization parquet full scan, main modifications as follows : 

1. SpecificOapRecordReaderBase extends our org.apache.parquet.hadoop.api.RecordReader not org.apache.parquet.hadoop.api.RecordReader

2. SpecificOapRecordReaderBase&VectorizedOapRecordReader initialize with configuration& footer by the constructor, not taskAttemptContext & inputSplit

3. add VectorizedColumnReaderWrapper let let readBatch method can be used outside "org.apache.spark.sql.execution.datasources.parquet" package

4. change some methods & parameters modifier to protected, then it can be access by subclass 

## How was this patch tested?

VectorizedOapEncodingSuite&VectorizedOapIOSuite

### CODE REVIEW

**Comment by linhong-intel** [comment]
one question, but not related to this question:
`OapReadSupportImpl` seems same to `ParquetReadSupport`
why not just use `ParquetReadSupport`

**Comment by LuciferYang** [comment]
@linhong-intel ur right, `OapReadSupportImpl` is same to `ParquetReadSupport` but `ParquetReadSupport` is private[parquet] access permission and `OapReadSupportImpl` proxy `ParquetReadSupport` use `ParquetReadSupportHelper`.

**Comment by linhong-intel** [comment]
Can you highlight your changes compared to `VectorizedParquetRecordReader` and `SpecificParquetRecordReaderBase`.

File: src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+     * rows of all the row groups.
+     */
+    protected long totalRowCount;
+    protected ParquetFileReader reader;
+
+    /**
+     * SpecificOapRecordReaderBase need
+     */
+    protected Configuration configuration;
+    protected ParquetMetadata footer;
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java]: configuration & footer use by initialize method, not belong to SpecificParquetRecordReaderBase

File: src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import org.apache.parquet.hadoop.metadata.BlockMetaData;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+import org.apache.parquet.hadoop.utils.Collections3;
+import org.apache.parquet.schema.MessageType;
+import org.apache.spark.sql.execution.datasources.oap.io.OapReadSupportImpl;
+import org.apache.spark.sql.execution.datasources.parquet.ParquetReadSupportHelper;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.StructType$;
+
+public abstract class SpecificOapRecordReaderBase<T> implements RecordReader<T> {
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java]: SpecificParquetRecordReaderBase extend org.apache.hadoop.mapreduce.RecordReader, SpecificOapRecordReaderBase extends our RecordReader

Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java]: SpecificParquetRecordReaderBase extends org.apache.hadoop.mapreduce.RecordReader and SpecificOapRecordReaderBase extends org.apache.parquet.hadoop.api.RecordReader

File: src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * SpecificOapRecordReaderBase need
+     */
+    protected Configuration configuration;
+    protected ParquetMetadata footer;
+
+
+    @Override
+    public void initialize() throws IOException, InterruptedException {
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java]: SpecificOapRecordReaderBase init method,needn't taskAttemptContext & inputSplit

**Pull request merged**